src/storage/src/render/persist_sink.rs

```diff
@@ -932,439 +935,574 @@ where
-                    batch_description_frontier: {:?}",
-                in_flight_descriptions,
-                done_batches,
-                batches_frontier,
-                batch_description_frontier
-            );
-
-            // Append batches in order, to ensure that their `lower` and
-            // `upper` line up.
-            done_batches.sort_by(|a, b| {
+            }
+            else => {
+                // All inputs are exhausted, so we can shut down.
+                return Ok(());
+            }
+        };
+        // Peel off any batches that are not beyond the frontier
+        // anymore.
+        //
+        // It is correct to consider batches that are not beyond the
+        // `batches_frontier` because it is held back by the writer
+        // operator as long as a) the `batch_description_frontier` did
+        // not advance and b) as long as the `desired_frontier` has not
+        // advanced to the `upper` of a given batch description.
+
+        let mut done_batches = in_flight_descriptions
+            .iter()
+            .filter(|(lower, _upper)| !PartialOrder::less_equal(&batches_frontier, lower))
+            .cloned()
+            .sorted_by(|a, b| {
                 if PartialOrder::less_than(a, b) {
                     Ordering::Less
                 } else if PartialOrder::less_than(b, a) {
                     Ordering::Greater
                 } else {
                     Ordering::Equal
                 }
-            });
+            })
+            .collect::<Vec<_>>();
+
+        trace!(
+            "persist_sink {collection_id}/{shard_id}: \
+                    append_batches: in_flight_descriptions size: {:?}, \
+                    in_flight_batches size: {:?}, \
+                    done size: {:?}, \
+                    batch_frontier: {:?}, \
+                    batch_description_frontier: {:?}",
+            in_flight_descriptions.len(),
+            in_flight_batches.len(),
+            done_batches.len(),
+            batches_frontier,
+            batch_description_frontier
+        );
 
-            // Reverse, as we'll pop batches off the end of the queue.
-            done_batches.reverse();
+        let result = update_persist_with_batches(
+            collection_id,
+            worker_id,
+            shard_id,
+            &metrics,
+            &busy_signal,
+            read_only_rx,
+            bail_on_concurrent_modification,
+            write,
+            upper_cap_set,
+            &current_upper,
+            &source_statistics,
+            &mut done_batches,
+            &mut in_flight_batches,
+            &mut in_flight_descriptions,
+        )
+        .await;
+
+        if let Err(e) = result {
+            return Err(e);
+        }
+    }
+}
 
-            while let Some(done_batch_metadata) = done_batches.pop() {
-                in_flight_descriptions.remove(&done_batch_metadata);
+async fn update_persist_with_batches(
```

Comment:
Does this need to be a separate function? Its logic only makes sense in the context of the function body above, since they also share a bunch of mutable state (`done_batches` and `in_flight_descriptions`). We should keep this as one function or try to come up slices of the logic that can stand on their own.